articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+| Mont Blanc                    |   off    | Fast-RTPS     |     1500     |   30    |   130    |
+| Mont Blanc                    | standard | Fast-RTPS     |      950     |   26    | 154->159 |
+| Mont Blanc                    |   new    | Fast-RTPS     |      220     |   14    |   130    |
+
+For what concerns latency and CPU usage, Sierra Nevada behaves almost the same regardless if standard IPC is enabled or not.
+This is due to the fact that most of its messages are very small in size.
+On the other hand, there are noticeable improvements in Mont Blanc, where several messages of non-negligible size are used.
+
+From the memory point of view, there is an almost constant increase in the utilization during the execution of the program when standard intra-process communication mechanism is used.
+Since the experiments have been run for 120 seconds, there is an increase of approximately 60KB per second.
+However, even considering the initial memory usage, it is possible to see how it is affected from the presence of the additional publishers and subscriptions needed for intra-process communication.
+There is a difference of 10MB in Sierra Nevada and of 33MB in Mont Blanc between standard intra-process communication on and off.
+
+The last experiment show how the current implementation performs in the case that both intra and inter-process communication are needed.
+The test consists of running Sierra Nevada on RaspberryPi 2, and, in a separate desktop machine, a single node subscribing to all the available topics coming from Sierra Nevada.
+This use-case is common when using tools such as `rosbag` or `rviz`.
+
+| ROS2 system                   |  IPC     |    RMW        | Latency [us] | CPU [%] | RAM [Mb] |
+| -------------                 |  -----   | ------------- | ------------ | ------- | -------- |
+| Sierra Nevada + debug node    |   off    | Fast-RTPS     |      800     |   22    |    50    |
+| Sierra Nevada + debug node    | standard | Fast-RTPS     |     1100     |   35    |  60->65  |
+| Sierra Nevada + debug node    |   new    | Fast-RTPS     |      180     |   15    |    32    |
+
+
+These results show that if there is at least one node in a different process, with the current implementation it is better to keep intra-process communication disabled.
+The proposed implementation does not require the ROS2 middleware when publishing intra-process.
+This allows to easily remove the connections between nodes in the same process when it is required to publish also inter process, potentially resulting in a very small overhead with respect to the only intra-process case.
+
+
+## Open Issues
+
+There are some open issues that are not addressed neither on the current implementation nor on the proposed one.
+
+ - The proposal does not take into account the problem of having a queue with twice the size when both inter and intra-process communication are used.
```

Comment:
I think that just documenting it would be enough.

Reply:
And in some cases, the duplication is just a buffer of "pointers".

Reply:
Yes, it is more a problem from a "conceptual" point of view rather than for the performances. 
In any case enforcing this would be quite complex as it requires constant communication between `rclcpp` and `rmw` layers.

Reply:
Yeah, I wouldn't enforce that. Just document the extra memory requirements.